crates/rattler/src/libsolv/pool.rs

```diff
@@ -0,0 +1,300 @@
+        Solver::new(solver)
+    }
+
+    /// Create the whatprovides on the pool which is needed for solving
+    pub fn create_whatprovides(&mut self) {
+        // Safe because pointer must exist
+        unsafe {
+            ffi::pool_createwhatprovides(self.as_ptr().as_mut());
+        }
+    }
+}
+
+/// Intern string like types
+fn intern_str<T: AsRef<str>>(pool: &mut PoolRef, str: T) -> StringId {
+    // Safe because conversion is valid
+    let c_str = CString::new(str.as_ref()).expect("could never be null because of trait-bound");
+    let length = c_str.as_bytes().len();
+    let c_str = c_str.as_c_str();
+
+    // Safe because pool exists and function accepts any string
+    unsafe {
+        StringId(ffi::pool_strn2id(
+            pool.as_ptr().as_mut(),
+            c_str.as_ptr(),
+            length.try_into().expect("string too large"),
+            1,
+        ))
+    }
+}
+
+/// Finds a previously interned string or returns `None` if it wasnt found.
+fn find_intern_str<T: AsRef<str>>(pool: &PoolRef, str: T) -> Option<StringId> {
+    // Safe because conversion is valid
+    let c_str = CString::new(str.as_ref()).expect("could never be null because of trait-bound");
+    let length = c_str.as_bytes().len();
+    let c_str = c_str.as_c_str();
+
+    // Safe because pool exists and function accepts any string
+    unsafe {
+        let id = ffi::pool_strn2id(
+            pool.as_ptr().as_ptr(),
+            c_str.as_ptr(),
+            length.try_into().expect("string too large"),
+            0,
+        );
+        if id == 0 {
+            None
+        } else {
+            Some(StringId(id))
+        }
+    }
+}
+
+/// Interns from Target type to Id
+pub trait Intern {
+    type Id;
+
+    /// Interns the type in the [`Pool`]
+    fn intern(&self, pool: &mut PoolRef) -> Self::Id;
+}
+
+pub trait FindInterned: Intern {
+    /// Finds a previously interned instance in the specified [`Pool`]
+    fn find_interned_id(&self, pool: &PoolRef) -> Option<Self::Id>;
+}
+
+/// Wrapper for the StringId of libsolv
+#[derive(Copy, Clone)]
+pub struct StringId(pub(super) ffi::Id);
+
+impl StringId {
+    /// Resolve to the interned type returns a string reference
+    pub fn resolve<'a>(&self, pool: &'a PoolRef) -> &'a str {
+        // Safe because the new-type wraps the ffi::id and cant be created otherwise
+        unsafe {
+            let c_str = ffi::pool_id2str(pool.as_ptr().as_ptr(), self.0);
+            CStr::from_ptr(c_str).to_str().expect("utf-8 parse error")
+        }
+    }
+}
+
+/// Intern implementation for string reference
+impl<'s> Intern for &'s str {
+    type Id = StringId;
+
+    fn intern(&self, pool: &mut PoolRef) -> Self::Id {
+        intern_str(pool, self)
+    }
+}
+
+impl<'s> FindInterned for &'s str {
+    fn find_interned_id(&self, pool: &PoolRef) -> Option<Self::Id> {
+        find_intern_str(pool, self)
+    }
+}
+
+/// Intern implementation for owned Strings
+impl<'s> Intern for &'s String {
```

Comment:
Why do you need the lifetime here? Can't you just implement this for `String`? Since the `intern` method already takes a reference to `Self` anyway?

Reply:
We've discussed this before and it didn't work, but I do not recall why.

Reply:
I tried removing the lifetime and that works fine.

I think previously we used a different approach with traits. But with the current setup no lifetime is required because the `intern` method already takes `self` by reference.